### PR TITLE
add deterministic training

### PR DESCRIPTION
**Pull request by cahity**
## Proposed changes
I implemented a function that provides necessary settings for deterministic usage of PyTorch, NumPy and random. I added a seed argument to the `train.py` file and called it.
Since augmentations use imgaug, which uses NumPy, DataLoader of PyTorch can not provide determinism. So, I also implemented a worker_init_fn to pass to DataLoaders when determinism is desired.

Postscriptum: I think the worker_init function can be given both in deterministic case and indeterministic case. In indeterministic case, it still helps with correcting the NumPy seeds. So, I did not put any checks there.

## Related issues
https://github.com/eriklindernoren/PyTorch-YOLOv3/issues/673

### CODE REVIEW

**Review by Flova — COMMENTED** [review]
Thanks for your PR. This seems quite usefull!

File: pytorchyolo/utils/utils.py [hunk 1]
```diff
@@ -9,6 +9,32 @@
...
+import random
+
+
+def provide_determinism(seed=42):
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+
+    torch.backends.cudnn.benchmark = False
```
[the remark this answers is not in the document]
Comment (cahity) [reply pytorchyolo/utils/utils.py:21]: AFAIK, yes, it may cause indeterminism when used since cuDNN benchmarks some algorithms and 'picks the best one'. This process is dependent on architecture, and it could even change in the same architecture due to benchmark noise.

Further info: https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-benchmarking

Note that this may reduce performance, so this feature might be disabled by default. For me, reproducibility is more important most of the time.

File: pytorchyolo/utils/utils.py [hunk 2]
```diff
@@ -9,6 +9,32 @@
...
+    torch.backends.cudnn.deterministic = True
+
+def worker_seed_np(worker_id):
```
Comment (cahity) [inline pytorchyolo/utils/utils.py:24]: Function name can be changed to something more convenient. It seems that I left the suffix '_np', which is not inclusive since random lib also gets seeded.

File: pytorchyolo/train.py [hunk 3]
```diff
@@ -72,9 +73,13 @@ def run():
     parser.add_argument("--conf_thres", type=float, default=0.1, help="Evaluation: Object confidence threshold")
     parser.add_argument("--nms_thres", type=float, default=0.5, help="Evaluation: IOU threshold for non-maximum suppression")
     parser.add_argument("--logdir", type=str, default="logs", help="Directory for training log files (e.g. for TensorBoard)")
+    parser.add_argument("--seed", type=int, default=42, help="Makes results reproducable. Set -1 to disable.")
```
Comment (Flova) [inline pytorchyolo/train.py:76]: Maybe disable the feature by default because of the possible performance penalties.

**Review by Flova — APPROVED** [review]
